src/modes/draw_polygon.js

```diff
@@ -58,29 +43,41 @@ module.exports = function(ctx) {
       ctx.ui.setActiveButton(Constants.types.POLYGON);
       this.on('mousemove', CommonSelectors.true, handleMouseMove);
       this.on('click', CommonSelectors.true, handleClick);
-      this.on('keyup', CommonSelectors.isEscapeKey, stopDrawingAndRemove);
-      this.on('keyup', CommonSelectors.isEnterKey, finish);
+      this.on('keyup', CommonSelectors.isEscapeKey, () => {
+        ctx.store.delete([polygon.id], { silent: true });
+        ctx.events.changeMode(Constants.modes.SIMPLE_SELECT);
+      });
+      this.on('keyup', CommonSelectors.isEnterKey, () => {
+        ctx.events.changeMode('simple_select', { featureIds: [polygon.id] });
+      });
     },
 
     stop: function() {
       doubleClickZoom.enable(ctx);
       ctx.ui.setActiveButton();
 
-      // If it's invalid, just destroy the thing
-      if (!polygon.isValid()) {
+      polygon.removeCoordinate(`0.${currentVertexPosition}`);
+      if (polygon.isValid() && ctx.store.get(polygon.id) !== undefined) {
+        ctx.map.fire(Constants.events.CREATE, {
+          features: [polygon.toGeoJSON()]
+        });
+      }
+      else if (ctx.store.get(polygon.id) !== undefined) {
         ctx.store.delete([polygon.id], { silent: true });
+        ctx.events.changeMode(Constants.modes.SIMPLE_SELECT, {}, { silent: true });
       }
```

Comment:
These conditionals could be simplified a little by early returning if the feature is undefined:

```js
if (ctx.store.get(polygon.id) === undefied) return;
if (polygon.isValid()) {
 ..
} else {
 ..
}
```